skrifa/src/outline/autohint/shape.rs

```diff
@@ -0,0 +1,87 @@
+//! Shaping support for autohinting.
+
+use raw::{tables::gsub::Gsub, TableProvider};
+
+use crate::{charmap::Charmap, collections::SmallVec, FontRef, GlyphId, MetadataProvider};
+
+/// Determines the fidelity with which we apply shaping in the
+/// autohinter.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub(crate) enum ShaperMode {
+    /// Characters are mapped to nominal glyph identifiers and layout tables
+    /// are not used for style coverage.
+    ///
+    /// This matches FreeType when HarfBuzz support is not enabled.
+    Nominal,
+    /// Simple substitutions are applied according to script rules and layout
+    /// tables are used to extend style coverage beyond the character map.
+    #[allow(unused)]
+    BestEffort,
+}
+
+#[derive(Copy, Clone, Default, Debug)]
+pub(crate) struct ShapedGlyph {
+    pub id: GlyphId,
+    pub y_offset: i32,
+}
+
+/// Arbitrarily chosen to cover our max input size plus some extra to account
+/// for expansion from multiple substitution tables.
+const SHAPED_CLUSTER_INLINE_SIZE: usize = 16;
+
+pub(crate) type ShapedCluster = SmallVec<ShapedGlyph, SHAPED_CLUSTER_INLINE_SIZE>;
```

Comment:
cluster is not so widely known outside of a very small circle, suggest adding a reference that explains what clusters are